[MediaStream] Cleanup getPhotoSettings and getPhotoCapabilities

eric-carlson opened the pull request:
#### 8607642c71c8943b9f85a77e2dbafcde87a79414
<pre>
[MediaStream] Cleanup getPhotoSettings and getPhotoCapabilities
<a href="https://bugs.webkit.org/show_bug.cgi?id=265497">https://bugs.webkit.org/show_bug.cgi?id=265497</a>
<a href="https://rdar.apple.com/118909014">rdar://118909014</a>

Reviewed by Jean-Yves Avenard.

Don&apos;t pass the DOMPromise outside of ImageCapture, have MediaStreamTrack and below
pass and return a NativePromise.

* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::getPhotoCapabilities):
(WebCore::ImageCapture::getPhotoSettings):
* Source/WebCore/Modules/mediastream/ImageCapture.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getPhotoCapabilities):
(WebCore::MediaStreamTrack::getPhotoSettings):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::getPhotoCapabilities):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::getPhotoCapabilities):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
(WebCore::PhotoCapabilitiesOrError::PhotoCapabilitiesOrError): Deleted.
(WebCore::PhotoCapabilitiesOrError::operator bool const): Deleted.
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::getPhotoCapabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::getPhotoCapabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::getPhotoCapabilities):
(WebKit::UserMediaCaptureManagerProxy::getPhotoCapabilities):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::getPhotoCapabilities):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::getPhotoCapabilities):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

Canonical link: <a href="https://commits.webkit.org/271314@main">https://commits.webkit.org/271314@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c73da2afb31e346324007a36b94dad24a839a14d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27976 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6614 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30506 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28471 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8609 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4005 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28243 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5370 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24031 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4624 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4807 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25038 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31195 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25582 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25474 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31062 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2974 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28873 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6346 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6715 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5297 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -81,35 +81,42 @@ void ImageCapture::takePhoto(PhotoSettings&& settings, DOMPromiseDeferred<IDLInt
 }
 
-void ImageCapture::getPhotoCapabilities(PhotoCapabilitiesPromise&& promise)
+void ImageCapture::getPhotoCapabilities(DOMPromiseDeferred<IDLDictionary<PhotoCapabilities>>&& promise)
```
why not keep the typedef ? a tad less long.

eric-carlson replied:
It is shorter, but it is only used in this one method

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -351,58 +351,70 @@ auto MediaStreamTrack::takePhoto(PhotoSettings&& settings) -> Ref<TakePhotoPromi
-            promise.resolve(WTFMove(*result.capabilities));
-        });
+    PhotoCapabilitiesPromise::Producer producer;
+    Ref<PhotoCapabilitiesPromise> promise = producer;
```
you can do `producer.promise()` now to not have the style checker complain

jyavenard replied:
You don't need the producer here. 

Return the PhotoCapabilitiesPromise in the whenSettled call.

eric-carlson replied:
Nice!

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -351,58 +351,70 @@ auto MediaStreamTrack::takePhoto(PhotoSettings&& settings) -> Ref<TakePhotoPromi
-                return;
-            }
+    PhotoSettingsPromise::Producer producer;
+    Ref<PhotoSettingsPromise> promise = producer;
```
same here, the producer isn't necessary.
return m_private->getPhotoSettings()->whenSettled(queue, [...] (auto&& result) { return PhotoSettinedPromise::createAndResolve...)} );

jyavenard commented on Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
```diff
@@ -189,9 +189,9 @@ const RealtimeMediaSourceCapabilities& MediaStreamTrackPrivate::capabilities() c
+Ref<RealtimeMediaSource::PhotoCapabilitiesNativePromise> MediaStreamTrackPrivate::getPhotoCapabilities()
 {
-    m_source->getPhotoCapabilities(WTFMove(completion));
+    return m_source->getPhotoCapabilities();
```
I like this much better.

jyavenard commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -647,12 +647,15 @@ void UserMediaCaptureManagerProxy::takePhoto(RealtimeMediaSourceIdentifier sourc
     }
 
-    handler(PhotoCapabilitiesOrError("Device not available"_s));
+    proxy->getPhotoCapabilities()->whenSettled(RunLoop::main(), [handler = WTFMove(handler)] (auto&& result) mutable {
```
` proxy->getPhotoCapabilities()->whenSettled(RunLoop::main(), WTFMove(handler));`

should work I think

eric-carlson replied:
Of course!

jyavenard commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
```diff
@@ -100,7 +100,7 @@ class UserMediaCaptureManagerProxy : public IPC::MessageReceiver {
     void takePhoto(WebCore::RealtimeMediaSourceIdentifier, WebCore::PhotoSettings&&, TakePhotoCallback&&);
 
-    using GetPhotoCapabilitiesCallback = CompletionHandler<void(WebCore::PhotoCapabilitiesOrError&&)>;
+    using GetPhotoCapabilitiesCallback = CompletionHandler<void(Expected<WebCore::PhotoCapabilities, String>&&)>;
```
you can also use the `PhotoCapabilitiesNativePromise::Result` type too. works in the IDL definition too.

eric-carlson replied:
I would need to use `WebCore::RealtimeMediaSource::PhotoCapabilitiesNativePromise::Result`, which I think makes it harder to understand so I will leave it as-is

jyavenard commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -81,35 +81,42 @@ void ImageCapture::takePhoto(PhotoSettings&& settings, DOMPromiseDeferred<IDLInt
-    }
+    m_track->getPhotoSettings()->whenSettled(RunLoop::main(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
+        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [this, promise = WTFMove(promise), result = WTFMove(result), protectedThis, identifier = WTFMove(identifier)] () mutable {
+            if (!result) {
```
you should be able to do directly `promise.settle(WTFMove(result));`

eric-carlson replied:
I could, but we need to check the result anyway because I want to log success or failure

cdumez commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -81,35 +81,42 @@ void ImageCapture::takePhoto(PhotoSettings&& settings, DOMPromiseDeferred<IDLInt
-        return;
-    }
+    m_track->getPhotoCapabilities()->whenSettled(RunLoop::main(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
+        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [this, promise = WTFMove(promise), result = WTFMove(result), protectedThis, identifier = WTFMove(identifier)] () mutable {
```
Capturing `protectedThis` is redundant here. The function you're calling is named queueTask*KeepingObjectAlive*

eric-carlson replied:
Indeed 😳

cdumez commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -81,35 +81,42 @@ void ImageCapture::takePhoto(PhotoSettings&& settings, DOMPromiseDeferred<IDLInt
-        return;
-    }
+    m_track->getPhotoSettings()->whenSettled(RunLoop::main(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
+        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [this, promise = WTFMove(promise), result = WTFMove(result), protectedThis, identifier = WTFMove(identifier)] () mutable {
```
ditto.

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -324,84 +324,72 @@ auto MediaStreamTrack::takePhoto(PhotoSettings&& settings) -> Ref<TakePhotoPromi
-    m_private->takePhoto(WTFMove(settings))->whenSettled(RunLoop::main(),
-        [protectedThis = Ref { *this }, producer = WTFMove(producer)] (auto&& result) mutable {
+    return m_private->takePhoto(WTFMove(settings))->whenSettled(RunLoop::main(),
+        [protectedThis = Ref { *this }] (auto&& result) mutable -> Ref<TakePhotoPromise> {
```
the compiler can't deduct this one ? `-> Ref<TakePhotoPromise>` shouldn't be needed

eric-carlson replied:
Changed

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -324,84 +324,72 @@ auto MediaStreamTrack::takePhoto(PhotoSettings&& settings) -> Ref<TakePhotoPromi
-                return;
-            }
+    return m_private->getPhotoCapabilities()->whenSettled(RunLoop::main(),
+        [protectedThis = Ref { *this }] (auto&& result) mutable -> Ref<PhotoCapabilitiesPromise> {
```
ditto

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -324,84 +324,72 @@ auto MediaStreamTrack::takePhoto(PhotoSettings&& settings) -> Ref<TakePhotoPromi
-                return;
-            }
+    return m_private->getPhotoSettings()->whenSettled(RunLoop::main(),
+        [protectedThis = Ref { *this }] (auto&& result) mutable -> Ref<PhotoSettingsPromise> {
```
ditto

eric-carlson committed:
[MediaStream] Cleanup getPhotoSettings and getPhotoCapabilities
https://bugs.webkit.org/show_bug.cgi?id=265497
rdar://118909014

Reviewed by Jean-Yves Avenard.

Don't pass the DOMPromise outside of ImageCapture, have MediaStreamTrack and below
pass and return a NativePromise.

* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::getPhotoCapabilities):
(WebCore::ImageCapture::getPhotoSettings):
* Source/WebCore/Modules/mediastream/ImageCapture.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getPhotoCapabilities):
(WebCore::MediaStreamTrack::getPhotoSettings):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::getPhotoCapabilities):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::getPhotoCapabilities):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
(WebCore::PhotoCapabilitiesOrError::PhotoCapabilitiesOrError): Deleted.
(WebCore::PhotoCapabilitiesOrError::operator bool const): Deleted.
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::getPhotoCapabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::getPhotoCapabilities):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::getPhotoCapabilities):
(WebKit::UserMediaCaptureManagerProxy::getPhotoCapabilities):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::getPhotoCapabilities):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::getPhotoCapabilities):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:

Canonical link: https://commits.webkit.org/271314@main

youennf commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -68,61 +68,66 @@ void ImageCapture::takePhoto(PhotoSettings&& settings, DOMPromiseDeferred<IDLInt
     auto identifier = LOGIDENTIFIER;
     ALWAYS_LOG(identifier);
 
-    if (m_track->readyState() == MediaStreamTrack::State::Ended) {
```
Is the spec requiring to reject synchronously as done before or asynchronously like done now ?

youennf commented on Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
```diff
@@ -257,12 +257,10 @@ auto MockRealtimeVideoSource::takePhotoInternal(PhotoSettings&&) -> Ref<TakePhot
-        completion({ *m_photoCapabilities });
-        return;
-    }
+    if (m_photoCapabilities)
```
I would tend to write it as if (!m_photoCapabilities)...
